feat(hostnqn): use allowed hosts with initiator's hostnqn

tiagolobocastro opened the pull request:
feat(hostnqn): use allowed hosts with auto-generated hostnqn within device uri

Uses the "initiator" node to restricted the nvmf targets to a specific node.
Ex: if the nexus that connects to a replica is on node "super-sayan" then the
replica is setup with allowed hosts: [nqn.2019-05.io.openebs:node-name:super-sayan]
The nexus is then created with replica uri as such:
nvmf://10.1.0.9:8420/nqn.2019-05.io.openebs:9fdebe7c-2fe3-4647-89c0-9f12576709c5?uuid=9fdebe7c-2f
e3-4647-89c0-9f12576709c5&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

feat(csi-hostnqn): use nexus uuid to generate hostnqn within device uri

Uses the nexus uuid to restricted the nvmf target access.
todo: use the csi nodename instead, which will be done once the PR
that pass this down the stack gets merged.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

fix: use hostnqn from dataplane registration

This allows us to use whatever the node is configured for.
TODO: update cli to report hostnqn?

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(hostnqn): use allowed hosts with auto-generated hostnqn within device uri

Uses the "initiator" node to restricted the nvmf targets to a specific node.
Ex: if the nexus that connects to a replica is on node "super-sayan" then the
replica is setup with allowed hosts: [nqn.2019-05.io.openebs:node-name:super-sayan]
The nexus is then created with replica uri as such:
nvmf://10.1.0.9:8420/nqn.2019-05.io.openebs:9fdebe7c-2fe3-4647-89c0-9f12576709c5?uuid=9fdebe7c-2f
e3-4647-89c0-9f12576709c5&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(csi-hostnqn): use nexus uuid to generate hostnqn within device uri

Uses the nexus uuid to restricted the nvmf target access.
todo: use the csi nodename instead, which will be done once the PR
that pass this down the stack gets merged.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on common/src/types/v0/transport/nvme_nqn.rs:
```diff
@@ -74,6 +79,14 @@ impl NvmeNqn {
         Self::Unique { uuid }
     }
+    /// Generate a Mayastor OpenEbs type name.
+    pub fn new_ebs_name(name: &str) -> Self {
```
nit: `new_org_name()` as this would be org type I believe?

tiagolobocastro replied:
yes that's better :+1:

Abhinandan-Purkait commented on control-plane/agents/src/bin/core/volume/specs.rs:
```diff
@@ -728,11 +731,14 @@ impl ResourceSpecsLocked {
             let mut replica = self.replica(&replica_state.uuid).await?;
-            match replica.share(registry, &replica_state.into()).await {
-                Ok(uri) => Ok(uri.into()),
+            let host_nqn = HostNqn::new_ebs_name(nexus_node.as_str());
```
Question: Is it suppposed to be the replica node rather than nexus node? From your example, `super-sayan` is where the replica is right?

Abhinandan-Purkait replied:
For multiple replicas on node `super-sayan-1`, `super-sayan-2`, `super-sayan-3`, and nexus on `super-sayan-1`, the replica uri's would be like, 

`nvmf://10.1.0.9:8420/nqn.2019-05.io.openebs:rep-2-uuid?uuid=rep-2-uuid&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan-2`, 

`nvmf://10.1.0.10:8420/nqn.2019-05.io.openebs:rep-3-uuid?uuid=rep-3-uuid&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan-3` 

respectively right?  Is my understanding correct?

tiagolobocastro replied:
No, the allowed-host is where the nexus lives. Because it's the nexus that connects to the replica.
So would be:

`nvmf://10.1.0.9:8420/nqn.2019-05.io.openebs:rep-2-uuid?uuid=rep-2-uuid&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan-1`

`nvmf://10.1.0.10:8420/nqn.2019-05.io.openebs:rep-3-uuid?uuid=rep-3-uuid&hostnqn=nqn.2019-05.io.openebs:node-name:super-sayan-1`

Abhinandan-Purkait replied:
Okay. Understood.

tiagolobocastro committed:
fix: use hostnqn from dataplane registration

This allows us to use whatever the node is configured for.
TODO: update cli to report hostnqn?

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors try

tiagolobocastro commented:
bors merge
